source/slang/slang-ir-specialize.cpp

```diff
@@ -211,7 +212,28 @@ struct SpecializationContext
         IRGeneric* genericVal,
         IRSpecialize* specializeInst)
     {
-        // First, we want to see if an existing specialization
+        // We need to fold the generic arguments here in order to uniquely identify
+        // which specializations need to be generated.
+        // The folding of the generic-arguments are deferred until the specialization
+        // step here, because the exact value of the generic-arguments can be unknown
+        // until the specialization. The exact values of the generic-arguments may come
+        // from the other modules, and they will be unknown until linking the modules.
+        //
+        UInt argCount = specializeInst->getArgCount();
+        for (UInt ii = 0; ii < argCount; ++ii)
+        {
+            IRInst* originalArg = specializeInst->getArg(ii);
+            IRInst* foldedArg = tryConstantFoldInst(module, originalArg);
+            if (foldedArg == originalArg)
+                continue;
+
+            IRBuilder builder(module);
+            builder.setInsertBefore(originalArg);
+            originalArg->replaceUsesWith(foldedArg);
```

Comment:
Note that `replaceUsesWith` may delete and replace `specailizeInst` with some existing inst.

The safest way to do this is:
```
specailizeInst = builder.replaceOperand(specailizeInst->getArgOperand(ii), foldedArg)
```

You probably need to add that `getAgOperand()` function to return an `IRuse` instead of an IRInst*.

Reply:
Thank you for the information.
I modified it to use IRUse.